### PR TITLE
Beasty blacksmith parameter shoot

### PR DESCRIPTION
**Pull request by BeastyBlacksmith**
haven't tested it yet, but I guess this is the functionality I am missing at the moment.
`DEFAULT_NLSOLVE_2` is not really necessary but it is more expressive this way imo, or we change the name of the second argument in `DEFAULT_NLSOVE`.

### CODE REVIEW

File: src/solve.jl [hunk 1]
```diff
@@ -24,6 +24,30 @@ function solve(prob::BVProblem, alg::Shooting; kwargs...)
...
 end
 
+function solve(prob::BVProblem, alg::ParameterShooting; kwargs...)
+    bc = prob.bc
+    u0 = deepcopy(prob.u0)
+    # Form a root finding function.
+    loss = function (resid, minimizer)
+        uEltype = eltype(minimizer)
+        tspan = (uEltype(prob.tspan[1]),uEltype(prob.tspan[2]))
+        tmp_prob = ODEProblem(prob.f,minimizer,tspan)
```
Comment (ChrisRackauckas) [inline src/solve.jl]: Need to pass the parameters

File: src/solve.jl [hunk 2]
```diff
@@ -24,6 +24,30 @@ function solve(prob::BVProblem, alg::Shooting; kwargs...)
     sol
 end
 
+function solve(prob::BVProblem, alg::ParameterShooting; kwargs...)
+    bc = prob.bc
+    u0 = deepcopy(prob.u0)
+    # Form a root finding function.
+    loss = function (resid, minimizer)
```
Comment (ChrisRackauckas) [inline src/solve.jl]: The minimizer is the parameters, not the initial condition, so it needs to be placed correctly.

**Comment by ChrisRackauckas** [comment]
I think that parameter shooting is better handled by the stuff in DiffEqParamEstim.jl (which does shooting) because those do minimizing instead of rootfinding, and with parameters/data there might not be a true minimum. It may be best to just add a BVP-specific loss function for `build_loss_objective`.

**Comment by BeastyBlacksmith** [comment]
Well, my experience with numerical minimizing is varying. Somehow the outcome is not very predictable to me, thats why I tend to trust the rootfinding algorithms more.
So this approach might not be suited in general, plus I am still not quite sure how to update the initial conditions once the parameters have changed ( do I need a callback for this? ).

**Comment by ChrisRackauckas** [comment]
>Well, my experience with numerical minimizing is varying. Somehow the outcome is not very predictable to me, thats why I tend to trust the rootfinding algorithms more.

Any rootfinding algorithm is the same as a minimization algorithm on the derivative. You can use trust-region Newton to do the minimization and will get a similar result, just without requiring the `f(x)=0` convergence which isn't guaranteed when parameters get involved. But I guess it can be kept as a separate thing, more related to data matching.

>I am still not quite sure how to update the initial conditions once the parameters have changed ( do I need a callback for this? ).

Make the minimizer array be both of them concatenated together, and split it into two views.

**Comment by BeastyBlacksmith** [comment]
> Make the minimizer array be both of them concatenated together, and split it into two views.

I mean to have the initial conditions be a function of the parameters, that function has to be passed to the `solve` routine.
